// @tencent/tea-component/src/popconfirm/PopConfirm.tsx
import React, { Fragment } from "react";
import classNames from "classnames";
import { BubbleProps, Bubble } from "../bubble";
import { StyledProps } from "../_type";
import { useDefault } from "../_util/use-default";

export interface PopConfirmProps extends StyledProps {
  /**
   * 使用受控模式来控制弹出气泡的显示，配合 `onVisibleChange` 方法使用
   */
  visible?: BubbleProps["visible"];

  /**
   * 处理 `visible` 变化的情况
   */
  onVisibleChange?: BubbleProps["onVisibleChange"];

  /**
   * 提示标题
   */
  title?: React.ReactNode;

  /**
   * 提示信息
   */
  message?: React.ReactNode;

  /**
   * 底部渲染内容
   * @docType React.ReactNode | ((close: () => void) => React.ReactNode)
   */
  footer?: React.ReactNode | ((close: () => void) => React.ReactNode);

  /**
   * 气泡放置的位置，将影响三角的朝向和位置
   * @default "top"
   */
  placement?: BubbleProps["placement"];

  /**
   * 需要被气泡包裹的内容
   */
  children?: React.ReactNode;

  /**
   * 是否在父容器滚动时关闭
   * @default false
   */
  closeOnScroll?: BubbleProps["closeOnScroll"];
}

export function PopConfirm({
  title,
  message,
  children,
  placement,
  footer,
  className,
  style,
  closeOnScroll,
  visible: _visible,
  onVisibleChange: _onVisibleChange,
}: PopConfirmProps) {
  const [visible, onVisibleChange] = useDefault(
    _visible,
    false,
    _onVisibleChange
  );

  const close = () => onVisibleChange(false);

  return (
    <Bubble
      visible={visible}
      onVisibleChange={onVisibleChange}
      closeOnScroll={closeOnScroll}
      className={classNames("tea-popconfirm", className)}
      placement={placement}
      content={
        <Fragment>
          <div className="tea-popconfirm__body">
            <div className="tea-media">
              <div className="tea-media__body">
                {title && (
                  <h3 className="tea-popconfirm__messagetitle">{title}</h3>
                )}
                {message && (
                  <div className="tea-popconfirm__messagetext">{message}</div>
                )}
              </div>
            </div>
          </div>
          {footer && (
            <div className="tea-popconfirm__footer">
              {typeof footer === "function" ? footer.call(null, close) : footer}
            </div>
          )}
        </Fragment>
      }
      trigger="click"
      style={style}
    >
      {children}
    </Bubble>
  );
}
