// @tencent/tea2-component/src/breadcrumb/Breadcrumb.tsx
import React from "react";
import classNames from "classnames";
import { isChildOfType } from "../_util/is-child-of-type";
import { StyledProps } from "../_type";
import { withStatics } from "../_util/with-statics";
import { useConfig } from "../_util/config-context";

export interface BreadcrumbProps extends StyledProps {
  /**
   * 导航内容
   */
  children?: React.ReactNode;

  /**
   * 较大字号的样式，可用于 Layout.Content 标题处
   */
  large?: boolean;
}

export const Breadcrumb = withStatics(
  function Breadcrumb({ large, children, className, style }: BreadcrumbProps) {
    const { classPrefix } = useConfig();
    return (
      <ol
        className={classNames(`${classPrefix}-breadcrumb`, className, {
          [`${classPrefix}-breadcrumb--fz-large`]: large,
        })}
        style={style}
      >
        {React.Children.map(children, (child, index) => {
          if (
            isChildOfType(child, BreadcrumbItem) &&
            index === React.Children.count(children) - 1
          ) {
            return React.cloneElement(child, {
              current: true,
            });
          }
          return child;
        })}
      </ol>
    );
  },
  {
    Item: BreadcrumbItem,
  }
);

interface BreadcrumbItemProps extends StyledProps {
  current?: boolean;
  children?: React.ReactNode;
}

export function BreadcrumbItem({
  current,
  children,
  className,
  style,
}: BreadcrumbItemProps) {
  const { classPrefix } = useConfig();

  if (typeof children === "string") {
    // eslint-disable-next-line no-param-reassign
    children = <span>{children}</span>;
  }

  return (
    <li
      className={classNames(`${classPrefix}-breadcrumb__item`, className)}
      style={style}
    >
      {React.isValidElement<StyledProps>(children)
        ? React.cloneElement(children, {
            className: classNames(
              children.props.className,
              `${classPrefix}-breadcrumb__item-title`,
              {
                "is-current": current,
              }
            ),
          })
        : children}
    </li>
  );
}
