// @tencent/tea-component/src/tag/Tag.tsx
import React from "react";
import classNames from "classnames";
import { Icon } from "../icon";
import { StyledProps } from "../_type";

export interface TagProps extends StyledProps {
  /**
   * 标签内容
   */
  children: React.ReactNode;

  /**
   * 关闭时回调
   *
   * **传递该参数时将展示关闭按钮**
   */
  onClose?: (event: React.MouseEvent) => void;

  /**
   * 颜色主题
   *
   * @default "default"
   */
  theme?: "default" | "primary" | "success" | "warning" | "error";

  /**
   * 是否为深色背景
   *
   * @default false
   */
  dark?: boolean;
}

export function Tag({
  children,
  onClose,
  className,
  theme,
  dark,
  style,
}: TagProps) {
  const hasCloseHandler = typeof onClose === "function";
  const classSuffix = dark ? "-dark" : "";

  return (
    <div
      className={classNames("tea-tag", className, {
        "tea-tag--edit": hasCloseHandler,
        [`is-blue${classSuffix}`]: theme === "primary",
        [`is-success${classSuffix}`]: theme === "success",
        [`is-warning${classSuffix}`]: theme === "warning",
        // @ts-ignore
        [`is-error${classSuffix}`]: theme === "danger" || theme === "error",
      })}
      style={style}
    >
      <span>{children}</span>
      {hasCloseHandler && <Icon link type="dismiss" onClick={onClose} />}
    </div>
  );
}
