// @tencent/tea-component/lib/table/addons/removeable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var icon_1 = require("../../icon");
var fallbackColumnKey = "__removeable_addon__";
var rowDisabled = null;
function removeable(_a) {
    var _b = _a.onRemove, onRemove = _b === void 0 ? function () { return null; } : _b, _c = _a.width, width = _c === void 0 ? 46 : _c;
    return {
        onInjectProps: function (props) {
            rowDisabled = props.rowDisabled || (function () { return false; });
            var columns = tslib_1.__spread(props.columns, [
                {
                    key: fallbackColumnKey,
                    header: "",
                    width: width,
                    render: function () { return null; },
                },
            ]);
            return tslib_1.__assign({}, props, { columns: columns });
        },
        onInjectColumn: function (previous) { return function (record, rowKey, recordIndex, column) {
            // 不是目标列
            if (column.key !== fallbackColumnKey) {
                return previous(record, rowKey, recordIndex, column);
            }
            var _a = previous(record, rowKey, recordIndex, column), preChildren = _a.children, props = _a.props, result = tslib_1.__rest(_a, ["children", "props"]);
            var children = preChildren;
            // 非表头
            if (recordIndex !== -1 && !rowDisabled(record)) {
                children = (react_1.default.createElement("div", null,
                    react_1.default.createElement("button", { type: "button", className: "tea-btn tea-btn--icon", onClick: function (event) { return onRemove(rowKey, { event: event, record: record }); } },
                        react_1.default.createElement(icon_1.Icon, { type: "dismiss" }))));
            }
            return tslib_1.__assign({}, result, { props: props, children: children });
        }; },
    };
}
exports.removeable = removeable;
