// @tencent/tea2-app/lib/core/i18n.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var i18next_1 = __importDefault(require("i18next"));
var react_i18next_1 = require("react-i18next");
var hash_string_1 = __importDefault(require("hash-string"));
/**
 * 用于 <Trans> 中插值
 */
function Slot(_a) {
    var content = _a.content;
    return react_1.default.createElement(react_1.default.Fragment, null, content);
}
// 是否已经初始化
var hasInited = false;
// 当前语言
var lng = window["VERSION"] || "zh";
// 从句子计算哈希一个 key 值，该算法需要和 scanner 保持一致
var hashKey = function (value) {
    return "k_" +
        ("0000" + hash_string_1.default(value.replace(/\s+/g, "")).toString(36)).slice(-7);
};
// 约定中语言名不规范
var _lngs = {
    zh: "zh",
    en: "en",
    jp: "ja",
    ko: "ko",
};
/**
 * 国际化包括的信息以及所需的工具
 */
exports.i18n = {
    /**
     * 当前用户的国际化语言，已知语言：
     *  - `zh` 中文
     *  - `en` 英文
     *  - `jp` 日语
     *  - `ko` 韩语
     */
    lng: lng,
    /**
     * 当前用户的国际化语言，已知语言：
     *  - `zh` 中文
     *  - `en` 英文
     *  - `jp` 日语
     *  - `ko` 韩语
     */
    lang: lng,
    /**
     * 当前用户所在站点
     *  - `1` 表示国内站；
     *  - `2` 表示国际站；
     *
     * @type {1 | 2}
     */
    site: 1,
    /**
     * 注册国家
     */
    country: {
        name: "CN",
        code: "86",
    },
    /**
     * 初始化当前语言的国际化配置
     */
    init: function (_a) {
        var _b;
        var translation = _a.translation;
        if (hasInited) {
            // @ts-ignore
            if (!PRODUCTION) {
                console.warn("你已经初始化过 i18n，请勿重复初始化");
            }
            return;
        }
        hasInited = true;
        i18next_1.default.use(react_i18next_1.reactI18nextModule).init({
            // 使用语言
            lng: _lngs[lng],
            // 英文版 fallback 到中文版，其它语言 fallback 到英文版
            fallbackLng: lng === "en" ? "zh" : "en",
            // 翻译资源
            resources: (_b = {},
                _b[_lngs[lng]] = { translation: translation },
                _b),
            ns: "translation",
            defaultNS: "translation",
            interpolation: {
                escapeValue: false,
            },
            react: {
                hashTransKey: hashKey,
            },
        });
    },
    /**
     * 标记翻译句子
     * 详细的标记说明，请参考 http://tapd.oa.com/tcp_access/markdown_wikis/0#1020399462009480783
     */
    t: function (sentence, options) {
        var key = hashKey(sentence);
        return i18next_1.default.t(key, __assign({}, (options || {}), { defaultValue: sentence }));
    },
    /**
     * 标记翻译组件
     * 详细的标记说明，请参考 http://tapd.oa.com/tcp_access/markdown_wikis/0#1020399462009480783
     */
    Trans: react_i18next_1.Trans,
    /**
     * 用于 <Trans> 中插值
     */
    Slot: Slot,
};
var loginInfo = window.LOGIN_INFO;
if (typeof loginInfo === "object" && loginInfo) {
    var area = loginInfo.area, countryCode = loginInfo.countryCode, countryName = loginInfo.countryName;
    exports.i18n.site = area;
    Object.assign(exports.i18n.country, {
        name: countryName,
        code: countryCode,
    });
}
exports.getI18NInstance = function () { return (hasInited ? i18next_1.default : null); };
/**
 * @internal 国际化容器，内部使用
 */
exports.I18NProvider = react_i18next_1.I18nextProvider;
