// @tencent/tea-component/lib/datepicker/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var use_outside_click_1 = require("../_util/use-outside-click");
function getYearMonthDate(date) {
    return {
        year: date.year(),
        month: date.month(),
        date: date.date(),
    };
}
exports.getYearMonthDate = getYearMonthDate;
exports.DatePickerTrigger = function (_a) {
    var childrenElementRef = _a.childrenElementRef, overlayElementRef = _a.overlayElementRef, visible = _a.visible, render = _a.render, onOpen = _a.onOpen, onClose = _a.onClose;
    var listen = use_outside_click_1.useOutsideClick([childrenElementRef, overlayElementRef]).listen;
    listen(function () {
        if (visible) {
            onClose();
        }
    });
    return render({
        overlayProps: {},
        childrenProps: {
            onClick: function (event) {
                event.stopPropagation();
                if (!visible) {
                    onOpen();
                }
                else {
                    onClose();
                }
            },
        },
    });
};
