// @tencent/tea2-component/lib/tagselect/TagSelectInput.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var merge_refs_1 = require("../_util/merge-refs");
var input_1 = require("../input");
exports.TagSelectInput = react_1.forwardRef(function TagSelectInput(_a, ref) {
    var value = _a.value, maxWidth = _a.maxWidth, placeholder = _a.placeholder, props = tslib_1.__rest(_a, ["value", "maxWidth", "placeholder"]);
    var inputRef = react_1.useRef(null);
    var mirrorRef = react_1.useRef(null);
    var _b = tslib_1.__read(react_1.useState(0), 2), width = _b[0], setWidth = _b[1];
    react_1.useEffect(function () {
        mirrorRef.current.innerText = value || placeholder;
        var width = mirrorRef.current.clientWidth;
        setWidth(width > maxWidth ? maxWidth : width);
    }, [maxWidth, placeholder, value]);
    return (react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement(input_1.Input, tslib_1.__assign({}, props, { value: value, placeholder: placeholder, style: { width: width + 6 }, ref: merge_refs_1.mergeRefs(ref, inputRef) })),
        react_1.default.createElement("span", { ref: mirrorRef, style: {
                position: "absolute",
                top: -9999,
                left: 0,
                whiteSpace: "pre",
                fontSize: 12,
            } })));
});
