// @tencent/tea2-component/lib/tagsearchbox/valueselect/SingleValueSelect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/* eslint-disable */
var react_1 = tslib_1.__importDefault(require("react"));
var i18n_1 = require("../../i18n");
var list_1 = require("../../list");
var dropdown_1 = require("../../dropdown");
var config_context_1 = require("../../_util/config-context");
var keys = {
    "8": "backspace",
    "9": "tab",
    "13": "enter",
    "37": "left",
    "38": "up",
    "39": "right",
    "40": "down",
};
var SingleValueSelect = /** @class */ (function (_super) {
    tslib_1.__extends(SingleValueSelect, _super);
    function SingleValueSelect(props) {
        var _this = _super.call(this, props) || this;
        // 父组件调用
        _this.handleKeyDown = function (keyCode) {
            if (!keys[keyCode])
                return;
            var onSelect = _this.props.onSelect;
            var select = _this.state.select;
            switch (keys[keyCode]) {
                case "enter":
                case "tab":
                    if (onSelect) {
                        onSelect(_this.getValue(select));
                    }
                    return false;
                case "up":
                    _this.move(-1);
                    break;
                case "down":
                    _this.move(1);
                    break;
            }
        };
        _this.move = function (step) {
            var select = _this.state.select;
            var values = _this.props.values;
            var list = values;
            if (list.length <= 0)
                return;
            _this.setState({ select: (select + step + list.length) % list.length });
        };
        _this.handleClick = function (e, index) {
            e.stopPropagation();
            if (_this.props.onSelect) {
                _this.props.onSelect(_this.getValue(index));
            }
        };
        var _a = _this.props, values = _a.values, inputValue = _a.inputValue, onSelect = _a.onSelect;
        var select = -1;
        values.forEach(function (item, index) {
            if (item.name === inputValue) {
                select = index;
            }
        });
        _this.state = {
            select: select,
        };
        return _this;
    }
    SingleValueSelect.prototype.componentDidMount = function () {
        var select = this.state.select;
        if (select < 0 && this.props.onSelect) {
            this.props.onSelect(this.getValue(select));
        }
    };
    SingleValueSelect.prototype.componentWillReceiveProps = function (nextProps) {
        var values = nextProps.values, inputValue = nextProps.inputValue;
        var list = values.map(function (item) { return item.name; });
        var select = list.indexOf(inputValue);
        this.setState({ select: select });
    };
    SingleValueSelect.prototype.getValue = function (select) {
        if (select < 0)
            return [];
        var _a = this.props, values = _a.values, inputValue = _a.inputValue;
        var list = values;
        if (select < list.length) {
            return [list[select]];
        }
        else {
            var select_1 = list.map(function (item) { return item.name; }).indexOf(inputValue);
            this.setState({ select: select_1 });
            if (select_1 < 0)
                return [];
            return [list[select_1]];
        }
    };
    SingleValueSelect.prototype.render = function () {
        var _this = this;
        var select = this.state.select;
        var _a = this.props, t = _a.t, classPrefix = _a.config.classPrefix, values = _a.values, offset = _a.offset, maxHeight = _a.maxHeight;
        var list = values.map(function (item, index) {
            return (react_1.default.createElement(list_1.List.Item, { key: index, selected: select === index, onClick: function (e) { return _this.handleClick(e, index); } },
                react_1.default.createElement("span", { title: item.name, style: item.style || {} }, item.name)));
        });
        if (list.length === 0) {
            list.push(react_1.default.createElement("li", { role: "presentation", key: 0, className: "is-disabled" },
                react_1.default.createElement("div", { className: classPrefix + "-option-list__item", role: "menuitem" }, t.emptyText)));
        }
        return (react_1.default.createElement(dropdown_1.DropdownBox, { style: { left: offset }, onClick: function (e) { return e.stopPropagation(); } },
            react_1.default.createElement(list_1.List, { type: "option", style: { maxHeight: maxHeight } }, list)));
    };
    SingleValueSelect = tslib_1.__decorate([
        config_context_1.withConfig,
        i18n_1.withTranslation
    ], SingleValueSelect);
    return SingleValueSelect;
}(react_1.default.Component));
exports.SingleValueSelect = SingleValueSelect;
