// @tencent/tea2-component/src/regionselect/RegionOption.tsx
import React, { useContext } from "react";
import classNames from "classnames";
import { RegionFlag } from "./RegionFlag";
import { useConfig } from "../_util/config-context";

export interface RegionOptionProps {
  /**
   * 地域值
   */
  value: string;

  /**
   * 地域显示文案
   */
  children?: React.ReactNode;

  /**
   * 已废弃
   * @deprecated
   */
  flag?: string;

  /**
   * 已废弃
   * @deprecated
   */
  country?: string;

  /**
   * 是否被选中
   *
   * **如在 `RegionSelect` 中使用，无需传值**
   */
  selected?: boolean;

  /**
   * 是否显示提醒小红点
   * @default false
   */
  dot?: boolean;

  /**
   * 被点击时回调
   */
  onClick?: React.MouseEventHandler;
}

export interface RegionOptionContext {
  inject(props: RegionOptionProps): RegionOptionProps;
}
export const RegionOptionContext = React.createContext<RegionOptionContext>(
  null
);

export function RegionOption(props: RegionOptionProps) {
  const { classPrefix } = useConfig();
  const context = useContext(RegionOptionContext);

  if (context) {
    props = context.inject(props); // eslint-disable-line no-param-reassign
  }

  const { children, dot, selected, onClick } = props;

  return (
    <div
      className={classNames(`${classPrefix}-region__item`, {
        "is-selected": selected,
      })}
      onClick={onClick}
    >
      <RegionFlag />
      <span className={`${classPrefix}-region__name`}>
        {children}
        {dot && <i className={`${classPrefix}-region__new`} />}
      </span>
    </div>
  );
}
