// @tencent/tea2-component/src/justify/Justify.tsx
import React from "react";
import classNames from "classnames";
import { StyledProps } from "../_type";
import { useConfig } from "../_util/config-context";

export interface JustifyProps extends StyledProps {
  /**
   * 是否顶部对齐
   * @default false
   */
  top?: boolean;

  /**
   * 是否底部对齐
   * @default false
   */
  bottom?: boolean;

  /**
   * 左侧内容
   */
  left?: React.ReactNode;

  /**
   * 右侧内容
   */
  right?: React.ReactNode;
}

export function Justify({
  left,
  right,
  top,
  bottom,
  className,
  style,
}: JustifyProps) {
  const { classPrefix } = useConfig();
  return (
    <div
      className={classNames(`${classPrefix}-justify-grid`, className, {
        [`${classPrefix}-justify-grid--start`]: top,
        [`${classPrefix}-justify-grid--end`]: bottom,
      })}
      style={style}
    >
      {left != null && (
        <div
          className={`${classPrefix}-justify-grid__col ${classPrefix}-justify-grid__col--left`}
        >
          {left}
        </div>
      )}
      {right != null && (
        <div
          className={`${classPrefix}-justify-grid__col ${classPrefix}-justify-grid__col--right`}
        >
          {right}
        </div>
      )}
    </div>
  );
}
