// @tencent/tea-component/src/inputadornment/InputAdornment.tsx
import React from "react";
import classNames from "classnames";
import { StyledProps } from "../_type";

export interface InputAdornmentProps extends StyledProps {
  /**
   * 前缀装饰
   */
  before?: React.ReactNode;

  /**
   * 后缀装饰
   */
  after?: React.ReactNode;

  /**
   * 被装饰内容
   */
  children?: React.ReactNode;

  /**
   * 展示类型
   * @default “default”
   */
  appearence?: "default" | "pure";
}

const wrapAdornmentAddon = (addon: React.ReactNode) => (
  <div className="tea-input-group__addon">{addon}</div>
);

const wrapAdornmentText = (text: string, pure: boolean) => (
  <span
    className={pure ? "tea-form__help-text--inline" : "tea-input-group__text"}
  >
    {text}
  </span>
);

export function InputAdornment({
  before,
  after,
  children,
  appearence,
  className,
  style,
}: InputAdornmentProps) {
  if (typeof before === "string") {
    // eslint-disable-next-line no-param-reassign
    before = wrapAdornmentText(before, false);
  }
  if (typeof after === "string") {
    // eslint-disable-next-line no-param-reassign
    after = wrapAdornmentText(after, appearence === "pure");
  }

  return (
    <div className={classNames("tea-input-group", className)} style={style}>
      {before && wrapAdornmentAddon(before)}
      {children}
      {after && wrapAdornmentAddon(after)}
    </div>
  );
}

InputAdornment.defaultLabelAlign = "middle";

/**
 * @deprecated
 *
 * 请使用 InputAdornment 代替
 */
export const InputAdorment = InputAdornment;
