// @tencent/tea-component/src/_chart/theme/tea.ts
import { Theme } from "../type";

export function getConfig(): Theme {
  return {
    color: [
      "#2d70f6",
      "#6fd598",
      "#f6ca54",
      "#eb6455",
      "#8b4bd2",
      "#5dbdca",
      "#94d553",
      "#f1a03f",
      "#d34a83",
      "#6360de",
      "#71c860",
      "#af40b1",
    ],
    padding: {
      left: 0,
      top: 10,
      bottom: 15,
      right: 5,
    },
    yAxis: {
      axisLine: {
        lineStyle: {
          color: "#BBB",
        },
      },
      axisTick: {
        lineStyle: {
          color: "#BBB",
        },
      },
      axisLabel: {
        color: "#BBB",
      },
      splitLine: {
        lineStyle: {
          color: ["#E5E5E5"],
        },
      },
    },
    xAxis: {
      axisLine: {
        lineStyle: {
          color: "#BBB",
        },
      },
      axisTick: {
        lineStyle: {
          color: "#BBB",
        },
      },
      axisLabel: {
        color: "#BBB",
      },
      splitLine: {
        lineStyle: {
          color: ["#E5E5E5"],
        },
      },
    },
    legend: {
      height: 20,
      rect: {
        width: 16,
        height: 5,
      },
      circle: {
        radius: 6,
      },
      square: {
        width: 10,
      },
      text: {
        fontSize: 12,
        color: "#2B2B2B",
        disabledColor: "#CECECE",
      },
    },
  };
}
