// @tencent/tea2-component/lib/modal/Modal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ModalMain_1 = require("./ModalMain");
var ModalConfirm_1 = require("./ModalConfirm");
exports.ModalConfirm = ModalConfirm_1.ModalConfirm;
var ModalAlert_1 = require("./ModalAlert");
exports.ModalAlert = ModalAlert_1.ModalAlert;
var ModalShow_1 = require("./ModalShow");
var with_statics_1 = require("../_util/with-statics");
exports.Modal = with_statics_1.withStatics(ModalMain_1.Modal, {
    /**
     * 渲染对话框的主要内容
     */
    Body: ModalMain_1.Modal.Body,
    /**
     * 渲染对话框的底部内容
     */
    Footer: ModalMain_1.Modal.Footer,
    /**
     * 对话框消息内容，可置于 ModalMain..Body 中
     */
    Message: ModalMain_1.Modal.Message,
    /**
     * API 的方式唤起一个确认对话框
     * @returns 异步返回布尔值，为 `true` 则表示用户确认，为 `false` 则表示用户取消
     */
    confirm: ModalConfirm_1.confirm,
    /**
     * API 的方式唤起一个对话框显示成功信息
     */
    success: ModalAlert_1.success,
    /**
     * API 的方式唤起一个对话框显示失败信息
     */
    error: ModalAlert_1.error,
    /**
     * API 的方式唤起一个对话框显示提醒信息
     */
    alert: ModalAlert_1.alert,
    /**
     * API 的方式唤起一个对话框
     */
    show: ModalShow_1.show,
});
