// @tencent/tea2-app/lib/core/history/create-history.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var warning_1 = __importDefault(require("warning"));
var history_1 = require("history");
var createTransitionManager_1 = __importDefault(require("history/createTransitionManager"));
/**
 * 利用控制台的 Router 对象创建兼容 React Router 的 history
 *
 * 参考 createBrowserHistory 实现
 * @see https://github.com/ReactTraining/history/blob/master/modules/createBrowserHistory.js
 */
function createHistory(props) {
    var globalHistory = window.history;
    var controller = props.controller, action = props.action, getUserConfirmation = props.getUserConfirmation, _a = props.keyLength, keyLength = _a === void 0 ? 6 : _a;
    var consoleRouter = window.seajs.require("router");
    var consoleRouteRule = "/" + controller + "/" + action + "(/*params)";
    var transitionManager = createTransitionManager_1.default();
    var initialLocation = getDOMLocation(getHistoryState());
    var history = {
        length: globalHistory.length,
        action: "POP",
        location: initialLocation,
        createHref: createHref,
        push: perform("PUSH"),
        replace: perform("REPLACE"),
        go: go,
        goBack: goBack,
        goForward: goForward,
        block: block,
        listen: listen,
    };
    function perform(action) {
        return function (path, state) {
            warning_1.default(!(typeof path === "object" &&
                path.state !== undefined &&
                state !== undefined), "You should avoid providing a 2nd state argument to " + action.toLocaleLowerCase() + " when the 1st " +
                "argument is a location-like object that already has state; it is ignored");
            var location = history_1.createLocation(path, state, createKey(), history.location);
            transitionManager.confirmTransitionTo(location, action, getUserConfirmation, function (ok) {
                if (!ok)
                    return;
                var href = createHref(location);
                var key = location.key, state = location.state;
                // createBrowserHistory 的实现是调用 history.pushState()
                // 这里改为通过控制台路由代理
                muteNextPop = true;
                consoleRouter.navigateWithState(href, false, action === "REPLACE", {
                    key: key,
                    state: state,
                });
                muteNextPop = false;
                setState({ action: action, location: location });
            });
        };
    }
    function go(n) {
        globalHistory.go(n);
    }
    function goBack() {
        go(-1);
    }
    function goForward() {
        go(1);
    }
    var listenerCount = 0;
    function checkRouterListeners(delta) {
        listenerCount += delta;
        if (listenerCount === 1 && delta === 1) {
            // add
            consoleRouter.use(consoleRouteRule, handlePopState);
        }
        else if (listenerCount === 0) {
            // remove
            consoleRouter.unuse(consoleRouteRule);
        }
    }
    function handlePopState() {
        var location = getDOMLocation(getHistoryState());
        handlePop(location);
    }
    var muteNextPop = false;
    function handlePop(location) {
        // push/replace 触发 console router.navigate() 方法
        // 同样会回调 handlePop()
        // 实际上，此时状态已经设置完毕，无需处理
        if (muteNextPop) {
            return;
        }
        var action = "POP";
        setState({ action: action, location: location });
    }
    var isBlocked = false;
    function block(prompt) {
        if (prompt === void 0) { prompt = false; }
        var unblock = transitionManager.setPrompt(prompt);
        if (!isBlocked) {
            checkRouterListeners(1);
            isBlocked = true;
        }
        return function () {
            if (isBlocked) {
                isBlocked = false;
                checkRouterListeners(-1);
            }
            return unblock();
        };
    }
    function listen(listener) {
        var unlisten = transitionManager.appendListener(listener);
        checkRouterListeners(1);
        return function () {
            checkRouterListeners(-1);
            unlisten();
        };
    }
    function createHref(location) {
        return history_1.createPath(location);
    }
    function createKey() {
        return Math.random()
            .toString(36)
            .substr(2, keyLength);
    }
    function setState(nextState) {
        if (nextState) {
            Object.assign(history, nextState);
        }
        history.length = globalHistory.length;
        transitionManager.notifyListeners(history.location, history.action);
    }
    return history;
}
exports.createHistory = createHistory;
function getHistoryState() {
    try {
        return window.history.state || {};
    }
    catch (e) {
        // IE 11 sometimes throws when accessing window.history.state
        // See https://github.com/ReactTraining/history/pull/289
        return {};
    }
}
function getDOMLocation(historyState) {
    var _a = historyState || {}, _b = _a.key, key = _b === void 0 ? undefined : _b, _c = _a.state, state = _c === void 0 ? undefined : _c;
    var _d = window.location, pathname = _d.pathname, search = _d.search, hash = _d.hash;
    var path = pathname + search + hash;
    return history_1.createLocation(path, state, key);
}
