// @tencent/tea2-app/lib/core/history/history-context.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var useContext = react_1.default.useContext, useEffect = react_1.default.useEffect;
// 使用 Context 提供 history，以及设置 conformation 的方法
exports.HistoryContext = react_1.default.createContext(null);
exports.HistoryContext.displayName = "Tea.History";
/**
 * 使用适配控制台的 history（React Hooks）
 *
 * 具体使用方法请参考 http://tapd.oa.com/QCloud_2015/markdown_wikis/view/#1010103951008571667
 *
 * @param getConfirmation (experimental) 路由切换前，让用户确认
 *
 * @example
```js
import React from 'react';
import { Router, Route, Link } from 'react-router-dom';
import { useHistory } from '@tea/app';
import { CvmDetailInstance } from './components';

export function CvmDetail() {
  // 此处为 React Hooks，只能在 Function Component 中使用
  const history = useHistory();

  return (
    <Router history={history}>
      <div>
        <Link to="/cvm/detail/ins-1q2w3e4r">ins-1q2w3e4r</Link>
        <Route path="/cvm/detail/:instanceId" component={CvmDetailInstance}></Route>
      </div>
    </Router>
  );
}
```
 */
exports.useHistory = function (getConfirmation) {
    var _a = useContext(exports.HistoryContext), history = _a.history, setConfirmation = _a.setConfirmation;
    // 设置提供的确认函数
    useEffect(function () {
        if (getConfirmation) {
            setConfirmation(getConfirmation);
        }
        return function () { return setConfirmation(null); };
    }, []);
    return history;
};
// TODO 转移 Ref
// dev 模式下最外层组件使用转移 Ref 的 HOC 导致异常
function withHistory(WrappedComponent) {
    function TeaHistoryHoc(props) {
        var history = exports.useHistory();
        return react_1.default.createElement(WrappedComponent, __assign({}, props, { history: history }));
    }
    return TeaHistoryHoc;
}
exports.withHistory = withHistory;
