// @tencent/tea-component/lib/layout/LayoutContent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var classnames_1 = tslib_1.__importDefault(require("classnames"));
var create_rocket_1 = require("../_util/create-rocket");
var icon_1 = require("../icon");
var justify_1 = require("../justify");
function LayoutContentHeader(_a) {
    var showBackButton = _a.showBackButton, _b = _a.onBackButtonClick, onBackButtonClick = _b === void 0 ? function () { return null; } : _b, title = _a.title, subtitle = _a.subtitle, children = _a.children, operation = _a.operation, className = _a.className, style = _a.style;
    return (react_1.default.createElement("div", { className: classnames_1.default("tea-layout__content-header", className), style: style },
        react_1.default.createElement("div", { className: "tea-layout__header-title" },
            react_1.default.createElement(justify_1.Justify, { left: react_1.default.createElement(react_1.default.Fragment, null,
                    showBackButton && (react_1.default.createElement("a", { href: "javascript:;", className: "tea-content__header-back", onClick: onBackButtonClick },
                        react_1.default.createElement(icon_1.Icon, { type: "btnback" }))),
                    title && react_1.default.createElement("h2", { className: "tea-h2" }, title),
                    children,
                    subtitle && (react_1.default.createElement("span", { className: "tea-layout__header-text" }, subtitle))), right: operation }))));
}
exports.LayoutContentHeader = LayoutContentHeader;
function LayoutContentBody(_a) {
    var full = _a.full, children = _a.children, className = _a.className, style = _a.style;
    return (react_1.default.createElement("div", { className: classnames_1.default("tea-layout__content-body", className), style: style },
        react_1.default.createElement("div", { className: classnames_1.default("tea-layout__content-body-inner", {
                "tea-layout__content-body-inner--full": full,
            }) }, children)));
}
exports.LayoutContentBody = LayoutContentBody;
var LayoutContentFooter = create_rocket_1.createRocket("LayoutContentFooter", "div.tea-layout__content-footer");
exports.LayoutContent = Object.assign(create_rocket_1.createRocket("LayoutContent", "main.tea-layout__content", "div.tea-layout__content-inner"), {
    Header: LayoutContentHeader,
    Body: LayoutContentBody,
    Footer: LayoutContentFooter,
});
