// @tencent/tea-component/lib/layout/Layout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var classnames_1 = tslib_1.__importDefault(require("classnames"));
var create_rocket_1 = require("../_util/create-rocket");
var LayoutContent_1 = require("./LayoutContent");
var is_child_of_type_1 = require("../_util/is-child-of-type");
var LayoutHeader = create_rocket_1.createRocket("LayoutHeader", "header.tea-layout__header");
var LayoutFooter = create_rocket_1.createRocket("LayoutFooter", "footer.tea-layout__footer");
var LayoutSider = create_rocket_1.createRocket("LayoutFooter", "nav.tea-layout__sidebar");
function LayoutBody(_a) {
    var className = _a.className, style = _a.style, children = _a.children;
    var _b = tslib_1.__read(react_1.useState(false), 2), hasSider = _b[0], setHasSider = _b[1];
    react_1.useEffect(function () {
        react_1.default.Children.forEach(children, function (child) {
            if (is_child_of_type_1.isChildOfType(child, LayoutSider)) {
                setHasSider(true);
            }
        });
    }, []); // eslint-disable-line react-hooks/exhaustive-deps
    return (react_1.default.createElement("section", { className: classnames_1.default("tea-layout__body", { "has-subsidebar": hasSider }, className), style: style }, children));
}
exports.Layout = Object.assign(create_rocket_1.createRocket("Layout", "section.tea-layout"), {
    Header: LayoutHeader,
    Body: LayoutBody,
    Footer: LayoutFooter,
    Sider: LayoutSider,
    Content: LayoutContent_1.LayoutContent,
});
