// @tencent/tea-component/lib/calendar/tables/DateTable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var moment_1 = tslib_1.__importStar(require("moment"));
var BaseTable_1 = tslib_1.__importStar(require("./BaseTable"));
var i18n_1 = require("../../i18n");
function DateTable(_a) {
    var value = _a.value, current = _a.current, _b = _a.onCurrentChange, onCurrentChange = _b === void 0 ? function () { return null; } : _b, range = _a.range, disabledDate = _a.disabledDate, _c = _a.onSelect, onSelect = _c === void 0 ? function () { return null; } : _c, _d = _a.onTypeChange, onTypeChange = _d === void 0 ? function () { return null; } : _d, _e = _a.cellStatus, cellStatus = _e === void 0 ? function () { return BaseTable_1.CellStatus.Common; } : _e;
    var t = i18n_1.useTranslation();
    var year = current.year();
    var month = current.month();
    function genTable() {
        var table = [];
        // 本月第一天
        var first = moment_1.default({
            year: year,
            month: month,
            date: 1,
        });
        var firstDay = first.day();
        // 第一周
        var firstWeek = Array(7)
            .fill(null)
            .map(function (_, i) { return first.clone().add(i - firstDay, "d"); });
        table.push(firstWeek);
        var nextWeekFirst = firstWeek[6].clone().add(1, "d");
        while (nextWeekFirst.month() === month) {
            var week = Array(7)
                .fill(null)
                .map(function (_, i) { return nextWeekFirst.clone().add(i, "d"); }); // eslint-disable-line no-loop-func
            table.push(week);
            nextWeekFirst = week[6].clone().add(1, "d");
        }
        return table;
    }
    function renderYear() {
        return (react_1.default.createElement("a", { className: "op-item", onClick: function () { return onTypeChange("year"); } }, current.format(t.yearFormat)));
    }
    function renderMonth() {
        return (react_1.default.createElement("a", { className: "op-item", onClick: function () { return onTypeChange("month"); } }, moment_1.default.monthsShort(month)));
    }
    var _f = tslib_1.__read(range || [null, null], 2), rangeMin = _f[0], rangeMax = _f[1];
    function isValid(date) {
        if (moment_1.isMoment(rangeMin) && rangeMin.isAfter(date, "day")) {
            return false;
        }
        if (moment_1.isMoment(rangeMax) && rangeMax.isBefore(date, "day")) {
            return false;
        }
        return disabledDate(date);
    }
    return (react_1.default.createElement(BaseTable_1.default, { caption: t.monthBeforeYear ? (react_1.default.createElement(react_1.default.Fragment, null,
            renderMonth(),
            renderYear())) : (react_1.default.createElement(react_1.default.Fragment, null,
            renderYear(),
            renderMonth())), jumperOptions: {
            onNext: function () { return onCurrentChange(moment_1.default(current).add(1, "month")); },
            onPrev: function () { return onCurrentChange(moment_1.default(current).subtract(1, "month")); },
            onCurrent: function () { return onCurrentChange(moment_1.default()); },
            isCurrent: current.isSame(moment_1.default(), "month"),
            nextDisabled: moment_1.isMoment(rangeMax) &&
                rangeMax.isBefore(moment_1.default({ year: year, month: month, date: 1 }).add(1, "month"), "day"),
            prevDisabled: moment_1.isMoment(rangeMin) &&
                rangeMin.isAfter(moment_1.default({ year: year, month: month, date: 1 }).subtract(1, "day"), "day"),
            nextTitle: t.nextMonth,
            prevTitle: t.prevMonth,
            curTitle: t.curMonth,
        } },
        react_1.default.createElement("div", { className: "tea-calendar__header-group" },
            react_1.default.createElement("div", { className: "tea-calendar__row" }, moment_1.default.weekdaysMin().map(function (name) { return (react_1.default.createElement(BaseTable_1.default.Cell, { key: name, name: name })); }))),
        genTable().map(function (week, i) { return (react_1.default.createElement(BaseTable_1.default.Row, { key: month + "-" + i }, week.map(function (date, j) { return (react_1.default.createElement(BaseTable_1.default.Cell, { key: month + "-" + i + "-" + j, name: date.date(), value: date, onSelect: onSelect, status: cellStatus(date), isNow: date.isSame(moment_1.default(), "day"), outside: date.month() !== month, disabled: !isValid(date), onOutsideSelect: function (date) {
                return onCurrentChange(moment_1.default({ year: date.year(), month: date.month() }));
            } })); }))); })));
}
exports.DateTable = DateTable;
