// @tencent/tea2-component/src/switch/_example/SwitchTreeExample.jsx
import React, { useState } from "react";
import { CheckTree } from "@tencent/tea-component/lib/checktree";
import { Switch } from "@tencent/tea-component/lib/switch";

export default function SwitchTreeExample() {
  const [selected, setSelected] = useState([]);
  const relations = {
    gz: "gd",
    sz: "gd",
    fs: "gd",
    by: "gz",
    th: "gz",
    ns: "sz",
    ft: "sz",
    sd: "fs",
    cc: "fs",
  };

  return (
    <CheckTree
      relations={relations}
      value={selected}
      onChange={value => setSelected(value)}
    >
      <Switch name="gd">广东省</Switch>
      <ul style={{ paddingLeft: 20 }}>
        <li style={{ paddingTop: 5 }}>
          <Switch name="gz">广州市</Switch>
          <ul style={{ paddingLeft: 20 }}>
            <li style={{ paddingTop: 5 }}>
              <Switch name="by">白云区</Switch>
            </li>
            <li style={{ paddingTop: 5 }}>
              <Switch name="th">天河区</Switch>
            </li>
          </ul>
        </li>
        <li style={{ paddingTop: 5 }}>
          <Switch name="sz">深圳市</Switch>
          <ul style={{ paddingLeft: 20 }}>
            <li style={{ paddingTop: 5 }}>
              <Switch name="ns">南山区</Switch>
            </li>
            <li style={{ paddingTop: 5 }}>
              <Switch name="ft">福田区</Switch>
            </li>
          </ul>
        </li>
        <li style={{ paddingTop: 5 }}>
          <Switch name="fs">佛山市</Switch>
          <ul style={{ paddingLeft: 20 }}>
            <li style={{ paddingTop: 5 }}>
              <Switch name="sd">顺德区</Switch>
            </li>
            <li style={{ paddingTop: 5 }}>
              <Switch name="cc">禅城区</Switch>
            </li>
          </ul>
        </li>
      </ul>
    </CheckTree>
  );
}
