// @tencent/tea-component/src/_chart/type/chart.ts
/**
 * 图表组件的通用属性
 */
export interface ChartProps {
  /**
   * 图表的宽度
   *
   * @default auto
   */
  width?: number;
  /**
   * 图表的高度
   *
   */
  height: number;
  /**
   * 图表的数据源
   *
   * @default []
   */
  dataSource: {
    [key: string]: string | number;
  }[];

  /**
   * X 轴与 Y 轴的数据映射。(X 轴属性名)*(Y 轴属性名)，属性名对应 dataSource 中的属性名
   */
  position: string;

  /**
   * 折线的粗细
   *
   * @default 3
   */
  size?: number;

  /**
   * 使用颜色来区分数据系列，可以自定义颜色
   */
  color?: string | Color;

  /**
   * 列定义
   */
  scale?: {
    [key: string]: Scale;
  };

  /**
   * 提示框
   * @default true
   */
  tooltip?: boolean | Tooltip;

  /**
   * 图例
   */
  legend?: boolean | Legend;
}

/**
 * 数据映射（颜色）
 */
export interface Color {
  /**
   * 数据映射的属性名
   */
  key: string;
  /**
   * 颜色列表
   * @default Tea主题
   */
  colors?: string[];
  /**
   * 格式化处理，优先级大于颜色列表
   */
  formatter?: (key: string, index: number, isLast: boolean) => string;
}

/**
 * 列定义
 */
export interface Scale {
  /**
   * 系列别名
   */
  alias?: string;
  /**
   * 数据的值域，值域的值尽量设置为 5 或 10 的整数倍
   *
   * @default auto
   */
  domain?: [number, number];
}

/**
 * 提示框
 */
export interface Tooltip {
  /**
   * 是否显示提示框
   */
  enable?: boolean;
  /**
   * 数据格式化
   */
  formatter?: (values: MetaData[], index: number) => MetaData[] | string;
}

/**
 * 图例
 */
export interface Legend {
  /**
   * 是否显示图例
   */
  enable?: boolean;
  /**
   * 图例标签格式化
   */
  formatter?: (label: string, index: number, isLast: boolean) => string;
}

/**
 * 坐标轴
 */
export interface Axis {
  enable?: boolean;
  formatter?: (
    label: string | number | { valueOf(): number },
    index: number
  ) => string;
}

/**
 * 折线图通用属性
 */
export interface RectProps extends ChartProps {
  /**
   *  X 轴
   * @default true
   */
  xAxis?: boolean | Axis;

  /**
   *  Y 轴
   * @default true
   */
  yAxis?: boolean | Axis;
}

export interface MetaData {
  value: number | string;
  color?: string;
  label?: string;
  title?: string;
  row?: { [key: string]: number | string };
}
