// @tencent/tea-component/src/tagsearchbox/_example/TagSearchBoxExample.jsx
import React, { useState } from "react";
import { TagSearchBox } from "@tea/component/tagsearchbox";

const values = [
  { key: "1", name: "value-1" },
  { key: "2", name: "value-2" },
  { key: "3", name: "value-3" },
  { key: "4", name: "value-4" },
  { key: "5", name: "value-5" },
  { key: "6", name: "value-6" },
  { key: "7", name: "value-7" },
  { key: "8", name: "value-8", style: { color: "red" } },
];

const attributes = [
  {
    type: "input",
    key: "1",
    name: "主机名",
    values,
  },
  {
    type: "single",
    key: "2",
    name: "云主机ID",
    values,
  },
  {
    type: "single",
    key: "3",
    name: "云主机状态",
    values: () => {
      return new Promise(resolve => {
        setTimeout(() => {
          resolve(values);
        }, 1000);
      });
    },
  },
  {
    type: ["multiple", { all: false }],
    key: "4",
    name: "所属项目",
    values,
  },
  {
    type: "multiple",
    key: "5",
    name: "地域",
    values: () => values,
  },
  {
    type: "multiple",
    key: "6",
    name: "可用区",
    values: () => {
      return new Promise(resolve => {
        setTimeout(() => {
          resolve(values);
        }, 1000);
      });
    },
  },
  {
    type: "multiple",
    key: "7",
    name: "空选项",
    values: () => {
      return new Promise(resolve => {
        setTimeout(() => {
          resolve([]);
        }, 1000);
      });
    },
  },
  {
    type: ["multiple", { emptyText: "请前往添加标签" }],
    key: "7",
    name: "自定义空选项",
    values: () => {
      return new Promise(resolve => {
        setTimeout(() => {
          resolve([]);
        }, 1000);
      });
    },
  },
  {
    type: "multiple",
    key: "8",
    name: "可重复属性",
    values,
    reusable: true,
  },
  {
    type: "multiple",
    key: "9",
    name: "test:",
    values,
  },
];

const defaultValue = [
  {
    attr: null,
    values: [{ name: "test-1" }, { name: "test-2" }],
  },
  {
    attr: {
      key: "2",
      name: "云主机ID",
    },
    values: [{ name: "value-1" }],
  },
  {
    attr: {
      key: "5",
      name: "地域",
    },
    values: [{ name: "value-1" }, { name: "value-2" }],
  },
];

// 受控组件
function ControlledDemo() {
  const [tags, setTags] = useState([
    {
      attr: { key: "2", name: "云主机ID" },
      values: [{ name: "value-1" }],
    },
  ]);

  function onChange(tags) {
    console.log(tags);
    setTags(tags);
  }

  return (
    <TagSearchBox
      // @ts-ignore
      attributes={attributes}
      minWidth={420}
      value={tags}
      onChange={onChange}
    />
  );
}

export default function TagSearchBoxExample() {
  return (
    <div>
      <TagSearchBox
        // @ts-ignore
        attributes={attributes}
        minWidth={420}
        defaultValue={defaultValue}
        onChange={tags => console.log(tags)}
      />
      <div style={{ marginTop: "16px" }}>
        <span>受控组件（value 和 onChange 需要配合使用）：</span>
        <ControlledDemo />
      </div>
    </div>
  );
}
