// @tencent/tea2-component/lib/navmenu/NavMenu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var classnames_1 = tslib_1.__importDefault(require("classnames"));
var with_statics_1 = require("../_util/with-statics");
var dropdown_1 = require("../dropdown");
var config_context_1 = require("../_util/config-context");
function NavMenuItem(_a) {
    var _b;
    var _c = _a.type, type = _c === void 0 ? "default" : _c, selected = _a.selected, onClick = _a.onClick, className = _a.className, style = _a.style, overlay = _a.overlay, children = _a.children;
    var classPrefix = config_context_1.useConfig().classPrefix;
    var wrapperProps = {
        className: classnames_1.default(className, (_b = {},
            _b[classPrefix + "-nav__service"] = type === "default",
            _b[classPrefix + "-nav__logo"] = type === "logo",
            _b[classPrefix + "-nav__dropdown"] = type === "dropdown",
            _b[classPrefix + "-nav__operation"] = type === "icon",
            _b["is-selected"] = selected,
            _b)),
        style: style,
        onClick: onClick,
    };
    if (type === "dropdown") {
        return (react_1.default.createElement("div", tslib_1.__assign({}, wrapperProps),
            react_1.default.createElement(dropdown_1.Dropdown, { trigger: "hover", button: children, clickClose: false, boxClassName: classPrefix + "-nav__dropdown-box", placementOffset: 0 }, overlay)));
    }
    if (type === "icon") {
        return (react_1.default.createElement("div", tslib_1.__assign({}, wrapperProps),
            react_1.default.createElement("a", { className: classPrefix + "-nav__operation-inner" }, children)));
    }
    if (type === "logo") {
        return (react_1.default.createElement("div", tslib_1.__assign({}, wrapperProps),
            react_1.default.createElement("a", null, children)));
    }
    // default
    if (typeof children === "string") {
        return (react_1.default.createElement("div", tslib_1.__assign({}, wrapperProps),
            react_1.default.createElement("a", null, children)));
    }
    return react_1.default.createElement("div", tslib_1.__assign({}, wrapperProps), children);
}
exports.NavMenuItem = NavMenuItem;
exports.NavMenu = with_statics_1.withStatics(function NavMenu(_a) {
    var left = _a.left, right = _a.right, className = _a.className, style = _a.style;
    var classPrefix = config_context_1.useConfig().classPrefix;
    return (react_1.default.createElement("nav", { className: classnames_1.default(classPrefix + "-nav", className), style: style },
        left,
        react_1.default.createElement("div", { className: classPrefix + "-nav__flex-end" }),
        right));
}, {
    Item: NavMenuItem,
});
