// @tencent/tea-component/lib/navmenu/NavMenu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var classnames_1 = tslib_1.__importDefault(require("classnames"));
var with_statics_1 = require("../_util/with-statics");
var dropdown_1 = require("../dropdown");
function NavMenuItem(_a) {
    var _b = _a.type, type = _b === void 0 ? "default" : _b, selected = _a.selected, onClick = _a.onClick, className = _a.className, style = _a.style, overlay = _a.overlay, children = _a.children;
    var wrapperProps = {
        className: classnames_1.default(className, {
            "tea-nav__service": type === "default",
            "tea-nav__logo": type === "logo",
            "tea-nav__dropdown": type === "dropdown",
            "tea-nav__operation": type === "icon",
            "is-selected": selected,
        }),
        style: style,
        onClick: onClick,
    };
    if (type === "dropdown") {
        return (react_1.default.createElement("div", tslib_1.__assign({}, wrapperProps),
            react_1.default.createElement(dropdown_1.Dropdown, { trigger: "hover", button: children, clickClose: false, boxClassName: "tea-nav__dropdown-box", placementOffset: 0 }, overlay)));
    }
    if (type === "icon") {
        return (react_1.default.createElement("div", tslib_1.__assign({}, wrapperProps),
            react_1.default.createElement("a", { className: "tea-nav__operation-inner", href: "javascript:;" }, children)));
    }
    if (type === "logo") {
        return (react_1.default.createElement("div", tslib_1.__assign({}, wrapperProps),
            react_1.default.createElement("a", { href: "javascript:;" }, children)));
    }
    // default
    if (typeof children === "string") {
        return (react_1.default.createElement("div", tslib_1.__assign({}, wrapperProps),
            react_1.default.createElement("a", { href: "javascript:;" }, children)));
    }
    return react_1.default.createElement("div", tslib_1.__assign({}, wrapperProps), children);
}
exports.NavMenuItem = NavMenuItem;
exports.NavMenu = with_statics_1.withStatics(function NavMenu(_a) {
    var left = _a.left, right = _a.right, className = _a.className, style = _a.style;
    return (react_1.default.createElement("nav", { className: classnames_1.default("tea-nav", className), style: style },
        left,
        react_1.default.createElement("div", { className: "tea-nav__flex-end" }),
        right));
}, {
    Item: NavMenuItem,
});
