// @tencent/tea-component/lib/table/addons/selectable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var warning_1 = tslib_1.__importDefault(require("warning"));
var classnames_1 = tslib_1.__importDefault(require("classnames"));
var checktree_1 = require("../../checktree");
var checkbox_1 = require("../../checkbox");
var get_row_key_from_record_key_1 = require("../util/get-row-key-from-record-key");
var check_1 = require("../../check");
var inject_props_if_target_not_existed_1 = require("../util/inject-props-if-target-not-existed");
function selectable(options) {
    var relations = options.relations, value = options.value, onChange = options.onChange, all = options.all, targetColumnKey = options.targetColumnKey, indent = options.indent, rowSelect = options.rowSelect, _a = options.width, width = _a === void 0 ? 26 : _a, _b = options.render, render = _b === void 0 ? function (x) { return x; } : _b, _c = options.shouldRecordExcludeFromAll, shouldRecordExcludeFromAll = _c === void 0 ? "disabled" : _c, _d = options.rowSelectable, rowSelectable = _d === void 0 ? function () { return true; } : _d;
    if (typeof indent !== "undefined") {
        warning_1.default(false, "Property `indent` is deprecated. Please use `indentable` addon instead.");
    }
    var fallbackColumnKey = "__selectable_addon__";
    var fallbackAllKey = "__selectable_all__";
    var allKey = typeof all === "string" ? all : fallbackAllKey;
    var rowDisabled = null;
    return {
        onInjectProps: function (props) {
            var columns = props.columns;
            rowDisabled = props.rowDisabled || (function () { return false; });
            columns = inject_props_if_target_not_existed_1.injectPropsIfTargetNotExisted(columns, targetColumnKey, {
                key: fallbackColumnKey,
                width: width,
                header: null,
                render: function () { return null; },
            });
            return tslib_1.__assign({}, props, { columns: columns });
        },
        onInjectColumn: function (previous) { return function (record, rowKey, recordIndex, column) {
            // 不是目标列
            if (column.key !== targetColumnKey && column.key !== fallbackColumnKey) {
                return previous(record, rowKey, recordIndex, column);
            }
            var _a = previous(record, rowKey, recordIndex, column), preChildren = _a.children, props = _a.props, result = tslib_1.__rest(_a, ["children", "props"]);
            var children = preChildren;
            // 表头
            if (recordIndex === -1) {
                if (all !== false) {
                    children = (react_1.default.createElement(checkbox_1.Checkbox, { name: allKey }, children || react_1.default.createElement(react_1.Fragment, null, "\u00A0")));
                }
            }
            // 记录行
            else {
                var paddingLeft = 0;
                var depth = 0;
                if (indent > 0) {
                    var node = rowKey;
                    while (relations && relations[node]) {
                        depth += 1;
                        node = relations[node];
                    }
                    paddingLeft = indent * depth;
                }
                var checkbox = (react_1.default.createElement(checkbox_1.Checkbox, { className: "tea-form-check--table-select", name: rowKey }, children || react_1.default.createElement(react_1.Fragment, null, "\u00A0")));
                var element = paddingLeft ? (react_1.default.createElement("div", { style: { paddingLeft: paddingLeft } }, checkbox)) : (checkbox);
                children = render(element, {
                    children: preChildren,
                    record: record,
                    depth: depth,
                    rowKey: rowKey,
                    recordIndex: recordIndex,
                    disabled: rowDisabled(record),
                });
            }
            return tslib_1.__assign({}, result, { props: props, children: children });
        }; },
        onInjectTable: function (renderBody) { return function (props) {
            var e_1, _a;
            var recordKey = props.recordKey, records = props.records, _b = props.rowDisabled, rowDisabled = _b === void 0 ? function () { return false; } : _b;
            var rowKey = get_row_key_from_record_key_1.getRowKeyfromRecordKey(recordKey);
            // 哪些应该从 all 里面排除
            var shouldRecordExcludeFromAllFinal = null;
            if (typeof shouldRecordExcludeFromAll === "function") {
                shouldRecordExcludeFromAllFinal = shouldRecordExcludeFromAll;
            }
            else if (shouldRecordExcludeFromAll === "disabled") {
                shouldRecordExcludeFromAllFinal = rowDisabled;
            }
            shouldRecordExcludeFromAllFinal =
                shouldRecordExcludeFromAllFinal || (function () { return false; });
            var disabledKeys = Array.from(new Set(tslib_1.__spread(Object.keys(relations || {}), records.map(rowKey)))).filter(function (key) {
                return !rowSelectable(key, {
                    parent: (relations || {})[key],
                    children: Object.entries(relations || {})
                        .filter(function (_a) {
                        var _b = tslib_1.__read(_a, 2), _ = _b[0], value = _b[1];
                        return value === key;
                    })
                        .map(function (_a) {
                        var _b = tslib_1.__read(_a, 1), key = _b[0];
                        return key;
                    }),
                });
            });
            // all keys
            var keys = Array.from(new Set(tslib_1.__spread(Object.keys(relations || {}), records
                .filter(function (x) { return !shouldRecordExcludeFromAllFinal(x); })
                .map(rowKey)
                .filter(function (x) {
                return shouldRecordExcludeFromAll === "disabled"
                    ? !disabledKeys.includes(x)
                    : true;
            }))));
            // keys without children
            var rootKeys = relations ? keys.filter(function (key) { return !relations[key]; }) : keys;
            var relationWithLeafs = tslib_1.__assign({}, (relations || {}));
            try {
                // all === false 时同样创建 relations，因为 CheckTree 需要 relations 识别子节点
                for (var rootKeys_1 = tslib_1.__values(rootKeys), rootKeys_1_1 = rootKeys_1.next(); !rootKeys_1_1.done; rootKeys_1_1 = rootKeys_1.next()) {
                    var key = rootKeys_1_1.value;
                    relationWithLeafs[key] = allKey;
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (rootKeys_1_1 && !rootKeys_1_1.done && (_a = rootKeys_1.return)) _a.call(rootKeys_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            // 只有全选 Check 可用时
            var hasRelations = !!Object.keys(relationWithLeafs).length;
            return (react_1.default.createElement(checktree_1.CheckTree, { relations: relationWithLeafs, value: value, onChange: hasRelations
                    ? function (value, context) {
                        return onChange(value, tslib_1.__assign({}, context, { record: records.find(function (record, index) {
                                return context.check.name === rowKey(record, index);
                            }), selectedRecords: records.filter(function (record, index) {
                                return value.includes(rowKey(record, index));
                            }) }));
                    }
                    : null, disabledNames: tslib_1.__spread(records.filter(function (x) { return rowDisabled(x); }).map(rowKey), disabledKeys) }, renderBody(props)));
        }; },
        onInjectRow: function (renderRow) { return function (record, rowKey, recordIndex, columns) {
            var _a = renderRow(record, rowKey, recordIndex, columns), prepends = _a.prepends, appends = _a.appends, preRow = _a.row;
            var row = preRow;
            // 支持整行选择
            row = (react_1.default.createElement(SelectWrapper, { key: rowKey, name: rowKey, rowSelect: rowSelect && !rowDisabled(record) }, row));
            return { prepends: prepends, row: row, appends: appends };
        }; },
    };
}
exports.selectable = selectable;
function SelectWrapper(_a) {
    var name = _a.name, rowSelect = _a.rowSelect, children = _a.children, props = tslib_1.__rest(_a, ["name", "rowSelect", "children"]);
    var context = react_1.useContext(check_1.CheckContext);
    if (context) {
        var checkProps_1 = context.inject({ type: "checkbox", name: name });
        var onChange_1 = checkProps_1.onChange, value_1 = checkProps_1.value;
        var rowSelectProps = {
            onClick: function (event) {
                // 事件合并
                if (typeof props.onClick === "function") {
                    props.onClick(event);
                }
                if (typeof children.props.onClick === "function") {
                    children.props.onClick(event);
                }
                return onChange_1(!value_1, { event: event, check: checkProps_1 });
            },
        };
        return react_1.default.cloneElement(children, tslib_1.__assign({}, props, { className: classnames_1.default(props.className, children.props.className, {
                "is-selected": !!value_1,
            }) }, (rowSelect ? rowSelectProps : {})));
    }
    return children;
}
