// @tencent/tea2-component/lib/list/List.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var classnames_1 = tslib_1.__importDefault(require("classnames"));
var create_rocket_1 = require("../_util/create-rocket");
var tooltip_1 = require("../tooltip");
var dropdown_1 = require("../dropdown");
var forward_ref_with_statics_1 = require("../_util/forward-ref-with-statics");
var config_context_1 = require("../_util/config-context");
var is_child_of_type_1 = require("../_util/is-child-of-type");
exports.ListItem = react_1.forwardRef(function ListItem(_a, ref) {
    var children = _a.children, className = _a.className, current = _a.current, disabled = _a.disabled, selected = _a.selected, onClick = _a.onClick, tooltip = _a.tooltip, props = tslib_1.__rest(_a, ["children", "className", "current", "disabled", "selected", "onClick", "tooltip"]);
    return (react_1.default.createElement(tooltip_1.Tooltip, { title: tooltip },
        react_1.default.createElement("li", tslib_1.__assign({}, props, { ref: ref, className: classnames_1.default(className, {
                "is-current": current,
                "is-disabled": disabled,
                "is-selected": selected,
            }), onClick: disabled ? null : onClick }), children)));
});
function SubMenu(_a) {
    var label = _a.label, children = _a.children, _b = _a.placement, placement = _b === void 0 ? "right" : _b, className = _a.className, props = tslib_1.__rest(_a, ["label", "children", "placement", "className"]);
    var classPrefix = config_context_1.useConfig().classPrefix;
    var dropdownStyle = { position: undefined };
    var _c = tslib_1.__read(placement.split("-"), 2), basePlacement = _c[0], placementModifier = _c[1];
    if (basePlacement === "left") {
        Object.assign(dropdownStyle, {
            left: "auto",
            right: "100%",
        });
    }
    if (placementModifier === "end") {
        Object.assign(dropdownStyle, {
            top: undefined,
            transform: "translateY(-100%)",
            marginTop: 6,
        });
    }
    return (react_1.default.createElement("li", tslib_1.__assign({ className: classnames_1.default(classPrefix + "-list__submenu", className) }, props),
        label,
        react_1.default.createElement(dropdown_1.DropdownBox, { style: dropdownStyle }, children)));
}
exports.SubMenu = SubMenu;
exports.List = forward_ref_with_statics_1.forwardRefWithStatics(function List(_a, ref) {
    var _b;
    var className = _a.className, style = _a.style, children = _a.children, type = _a.type, split = _a.split, _c = _a.onScrollBottom, onScrollBottom = _c === void 0 ? function () { return null; } : _c;
    var classPrefix = config_context_1.useConfig().classPrefix;
    var _d = tslib_1.__read(react_1.useState(false), 2), hasSubMenu = _d[0], setHasSubMenu = _d[1];
    react_1.useEffect(function () {
        react_1.default.Children.forEach(children, function (child) {
            if (is_child_of_type_1.isChildOfType(child, SubMenu)) {
                setHasSubMenu(true);
            }
        });
    }, []); // eslint-disable-line react-hooks/exhaustive-deps
    function handleBodyScroll(event) {
        var list = event.target;
        var scrollHeight = list.scrollHeight, scrollTop = list.scrollTop, clientHeight = list.clientHeight;
        if (scrollHeight <= Math.round(clientHeight + scrollTop)) {
            onScrollBottom(event);
        }
    }
    var digits = String(react_1.default.Children.count(children)).length;
    var listClassName = classnames_1.default((_b = {},
        _b[classPrefix + "-list"] = true,
        _b[classPrefix + "-list--bullet"] = type === "bullet",
        _b[classPrefix + "-list--number"] = type === "number",
        _b[classPrefix + "-list--option"] = type === "option" || type === "option-group",
        _b[classPrefix + "-list--group"] = type === "option-group",
        _b[classPrefix + "-list--divider"] = split === "divide",
        _b[classPrefix + "-list--striped"] = split === "stripe",
        _b["is-2digits"] = type === "number" && digits === 2,
        _b["is-3digits"] = type === "number" && digits === 3,
        _b["is-4digits"] = type === "number" && digits > 3,
        _b[className] = className,
        _b));
    var Parent = type === "number" ? "ol" : "ul";
    return (react_1.default.createElement(Parent, { ref: ref, className: listClassName, style: tslib_1.__assign({ 
            // 多级菜单展开
            maxHeight: hasSubMenu ? "none" : undefined }, (style || {})), onScroll: handleBodyScroll }, children));
}, {
    GroupLabel: create_rocket_1.createRocket("GroupLabel", "li.@{prefix}-list__label"),
    StatusTip: create_rocket_1.createRocket("TipItem", "li.@{prefix}-list__status"),
    SubMenu: SubMenu,
    Item: exports.ListItem,
});
