// @tencent/tea2-component/src/tagsearchbox/_example/TagSearchBoxExample.jsx
import React, { useState } from "react";
import { TagSearchBox } from "@tencent/tea-component/lib/tagsearchbox";
import { List } from "@tencent/tea-component/lib/list";

const values = [
  { key: "1", name: "value-1" },
  { key: "2", name: "value-2" },
  { key: "3", name: "value-3" },
  { key: "4", name: "value-4" },
  { key: "5", name: "value-5" },
  { key: "6", name: "value-6" },
  { key: "7", name: "value-7" },
  { key: "8", name: "value-8", style: { color: "red" } },
  { key: "9", name: "value-9" },
  { key: "10", name: "value-10" },
  { key: "11", name: "value-11" },
  { key: "12", name: "value-12" },
  { key: "13", name: "value-13" },
  { key: "14", name: "value-14" },
  { key: "15", name: "value-15" },
];

/**
 * @type {import("@tencent/tea-component/lib/tagsearchbox").TagSearchBoxProps["attributes"]}
 */
const attributes = [
  {
    type: "input",
    key: "1",
    name: "主机名",
  },
  {
    type: "single",
    key: "2",
    name: "云主机ID",
    values,
    removeable: false,
  },
  {
    type: "single",
    key: "3",
    name: "云主机状态",
    values: () => {
      return new Promise(resolve => {
        setTimeout(() => {
          resolve(values);
        }, 1000);
      });
    },
  },
  {
    type: ["multiple", { all: false }],
    key: "4",
    name: "所属项目",
    values,
  },
  {
    type: "multiple",
    key: "5",
    name: "地域",
    values: () => values,
  },
  {
    type: "multiple",
    key: "6",
    name: "可用区",
    values: () => {
      return new Promise(resolve => {
        setTimeout(() => {
          resolve(values);
        }, 1000);
      });
    },
  },
  {
    key: "7",
    type: "render",
    name: "自定义渲染",
    render: ({ onSelect }) => {
      return (
        <>
          <List type="option">
            <List.Item onClick={() => onSelect([{ key: "1", name: "预付费" }])}>
              预付费
            </List.Item>
            <List.Item onClick={() => onSelect([{ key: "2", name: "后付费" }])}>
              后付费
            </List.Item>
            <List.SubMenu label="更多">
              <List type="option">
                <List.Item disabled>订阅</List.Item>
                <List.Item
                  onClick={() => onSelect([{ key: "3", name: "朋友代付" }])}
                >
                  朋友代付
                </List.Item>
              </List>
            </List.SubMenu>
          </List>
        </>
      );
    },
  },
  {
    type: "multiple",
    key: "8",
    name: "空选项",
    values: () => {
      return new Promise(resolve => {
        setTimeout(() => {
          resolve([]);
        }, 1000);
      });
    },
  },
  {
    type: ["multiple", { emptyText: "请前往添加标签" }],
    key: "9",
    name: "自定义空选项",
    values: () => {
      return new Promise(resolve => {
        setTimeout(() => {
          resolve([]);
        }, 1000);
      });
    },
  },
  {
    type: "multiple",
    key: "10",
    name: "可重复属性",
    values,
    reusable: true,
  },
  {
    type: "multiple",
    key: "11",
    name: "结尾冒号测试:",
    values,
  },
];

const defaultValue = [
  {
    attr: attributes[1],
    values: [attributes[1].values[1]],
  },
  {
    attr: null,
    values: [
      { key: "1", name: "test-1" },
      { key: "2", name: "test-2" },
    ],
  },
];

// 受控组件
function ControlledDemo() {
  const [tags, setTags] = useState([
    {
      attr: attributes[1],
      values: [{ key: "1", name: "value-1" }],
    },
  ]);

  function onChange(tags) {
    console.log(tags);
    setTags(tags);
  }

  return (
    <TagSearchBox
      attributes={attributes}
      minWidth={420}
      value={tags}
      onChange={onChange}
    />
  );
}

export default function TagSearchBoxExample() {
  return (
    <div>
      <TagSearchBox
        attributes={attributes}
        minWidth={420}
        defaultValue={defaultValue}
        onChange={tags => console.log(tags)}
      />
      <div style={{ marginTop: "16px" }}>
        <span>受控组件（value 和 onChange 需要配合使用）：</span>
        <ControlledDemo />
      </div>
    </div>
  );
}
