// @tencent/tea2-app/lib/bridge/user/user.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var current_1 = require("./current");
var login_1 = require("./login");
var csrf_1 = require("./csrf");
var whitelist_1 = require("./whitelist");
var region_1 = require("./region");
var project_1 = require("./project");
var event_1 = require("./event");
var userInfo = (window.LOGIN_INFO || {});
var isOwner = userInfo.loginUin && userInfo.loginUin === userInfo.ownerUin;
exports.user = Object.assign(event_1.userEmitter, __assign({}, userInfo, { isOwner: isOwner,
    current: current_1.current,
    getAntiCSRFToken: csrf_1.getAntiCSRFToken,
    checkWhitelist: whitelist_1.checkWhitelist,
    checkWhitelistBatch: whitelist_1.checkWhitelistBatch,
    getLastRegionId: region_1.getLastRegionId,
    setLastRegionId: region_1.setLastRegionId,
    getLastProjectId: project_1.getLastProjectId,
    setLastProjectId: project_1.setLastProjectId,
    getPermitedProjectInfo: project_1.getPermitedProjectInfo,
    getPermitedProjectList: project_1.getPermitedProjectList,
    login: login_1.login,
    logout: login_1.logout }));
