// @tencent/tea2-component/lib/domref/DomRef.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var react_dom_1 = tslib_1.__importDefault(require("react-dom"));
var merge_refs_1 = require("../_util/merge-refs");
var call_both_1 = require("../_util/call-both");
var DomRefInner = /** @class */ (function (_super) {
    tslib_1.__extends(DomRefInner, _super);
    function DomRefInner() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DomRefInner.prototype.componentDidMount = function () {
        var dom = react_dom_1.default.findDOMNode(this); // eslint-disable-line react/no-find-dom-node
        var domRef = this.props.domRef;
        merge_refs_1.mergeRefs(domRef)(dom);
    };
    DomRefInner.prototype.componentDidUpdate = function () {
        var dom = react_dom_1.default.findDOMNode(this); // eslint-disable-line react/no-find-dom-node
        var domRef = this.props.domRef;
        merge_refs_1.mergeRefs(domRef)(dom);
    };
    DomRefInner.prototype.componentWillUnmount = function () {
        var domRef = this.props.domRef;
        merge_refs_1.mergeRefs(domRef)(null);
    };
    DomRefInner.prototype.render = function () {
        var e_1, _a;
        var _b = this.props, children = _b.children, domRef = _b.domRef, props = tslib_1.__rest(_b, ["children", "domRef"]);
        var childrenElement = children;
        var attachProps = {};
        // 元素类型的，可以把要求渲染的 childrenProps 和原本的 props 合并
        if (react_1.default.isValidElement(childrenElement)) {
            try {
                for (var _c = tslib_1.__values(Object.entries(props)), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var _e = tslib_1.__read(_d.value, 2), propName = _e[0], propValue = _e[1];
                    if (typeof propValue === "function") {
                        // 子节点之前已经有处理函数，则同时调用
                        if (childrenElement.props[propName]) {
                            attachProps[propName] = call_both_1.callBoth(propValue, childrenElement.props[propName]);
                        }
                        else {
                            attachProps[propName] = propValue;
                        }
                    }
                    else {
                        attachProps[propName] = propValue;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return react_1.default.cloneElement(childrenElement, attachProps);
        }
        return children;
    };
    return DomRefInner;
}(react_1.default.Component));
exports.DomRef = react_1.forwardRef(function (props, ref) { return (react_1.default.createElement(DomRefInner, tslib_1.__assign({}, props, { domRef: ref }))); });
