// @tencent/tea2-component/lib/upload/Upload.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var react_dropzone_1 = require("react-dropzone");
var uploadFile_1 = require("./uploadFile");
var uuid_1 = require("../_util/uuid");
var inject_value_1 = require("../_util/inject-value");
var is_callable_1 = require("../_util/is-callable");
var File_1 = require("./preset/File");
var Dragger_1 = require("./preset/Dragger");
var Image_1 = require("./preset/Image");
var with_statics_1 = require("../_util/with-statics");
var UploadContext_1 = require("./UploadContext");
function isPromise(p) {
    return p && typeof p === "object" && typeof p.then === "function";
}
exports.Upload = with_statics_1.withStatics(function Upload(_a) {
    var action = _a.action, accept = _a.accept, maxSize = _a.maxSize, _b = _a.multiple, multiple = _b === void 0 ? false : _b, headers = _a.headers, data = _a.data, _c = _a.name, name = _c === void 0 ? "file" : _c, withCredentials = _a.withCredentials, _d = _a.beforeUpload, beforeUpload = _d === void 0 ? function (_, __, isAccepted) { return isAccepted; } : _d, _e = _a.onStart, onStart = _e === void 0 ? function () { return null; } : _e, onProgress = _a.onProgress, onSuccess = _a.onSuccess, onError = _a.onError, children = _a.children, className = _a.className, style = _a.style;
    var _f = tslib_1.__read(react_1.useState(false), 2), dragging = _f[0], setDragging = _f[1];
    var _g = react_dropzone_1.useDropzone({
        accept: accept,
        maxSize: maxSize,
        multiple: multiple,
        noClick: true,
        noKeyboard: true,
        onDrop: uploadFiles,
        onDragEnter: function () { return setDragging(true); },
        onDragLeave: function () { return setDragging(false); },
    }), getInputProps = _g.getInputProps, getRootProps = _g.getRootProps, open = _g.open;
    function uploadFiles(acceptedFiles, rejectedFiles) {
        var fileList = tslib_1.__spread(acceptedFiles, rejectedFiles);
        acceptedFiles.forEach(function (file) {
            file["id"] = uuid_1.uuid(); // eslint-disable-line dot-notation
            uploadFile(file, fileList, true);
        });
        rejectedFiles.forEach(function (file) {
            file["id"] = uuid_1.uuid(); // eslint-disable-line dot-notation
            uploadFile(file, fileList, false);
        });
        setDragging(false);
    }
    function uploadFile(file, fileList, isAccepted) {
        var couldUpload = beforeUpload(file, fileList, isAccepted);
        if (isPromise(couldUpload)) {
            couldUpload.then(function () { return request(file); }).catch(function () { return null; });
        }
        else {
            if (!couldUpload) {
                return;
            }
            setTimeout(function () { return request(file); }, 0);
        }
    }
    function request(file) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var uploadAction, xhr;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(typeof action === "function")) return [3 /*break*/, 2];
                        return [4 /*yield*/, action(file)];
                    case 1:
                        uploadAction = _a.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        uploadAction = action;
                        _a.label = 3;
                    case 3:
                        xhr = uploadFile_1.upload({
                            action: uploadAction,
                            filename: name,
                            file: file,
                            data: inject_value_1.injectValue(data)(file),
                            headers: inject_value_1.injectValue(headers)(file),
                            withCredentials: withCredentials,
                            onProgress: onProgress,
                            onSuccess: onSuccess,
                            onError: onError,
                        });
                        onStart(file, { xhr: xhr });
                        return [2 /*return*/];
                }
            });
        });
    }
    return (react_1.default.createElement(UploadContext_1.UploadContext.Provider, { value: { open: open, getDraggerProps: getRootProps, isDragging: dragging } },
        react_1.default.createElement("span", { className: className, style: style },
            react_1.default.createElement("input", tslib_1.__assign({}, getInputProps())),
            is_callable_1.isCallable(children) ? (children({
                open: open,
                getDraggerProps: getRootProps,
                isDragging: dragging,
            })) : (react_1.default.createElement("span", { onClick: open }, children)))));
}, {
    defaultLabelAlign: "middle",
    File: File_1.File,
    Dragger: Dragger_1.Dragger,
    Image: Image_1.Image,
});
