// @tencent/tea-component/src/table/_example/TableAddonExample.jsx
import React, { useState } from "react";
import { Table } from "@tea/component/table";
import { Icon } from "@tea/component/icon";

const {
  sortable,
  filterable,
  scrollable,
  radioable,
  injectable,
} = Table.addons;

const records = [
  { name: "张三", age: 8, stage: "teenager", marrage: 0 },
  { name: "李四", age: 18, stage: "youth", marrage: 0 },
  { name: "王五", age: 28, stage: "youth", marrage: 1 },
  { name: "赵六", age: 38, stage: "youth", marrage: 0 },
  { name: "孙七", age: 48, stage: "middle-aged", marrage: 2 },
  { name: "周八", age: 58, stage: "middle-aged", marrage: 1 },
  { name: "吴九", age: 68, stage: "elder", marrage: 3 },
  { name: "郑十", age: 78, stage: "elder", marrage: 4 },
];
const MARRAGE_TEXT = ["未婚", "已婚", "离异", "再婚", "丧偶"];
const STAGE_TEXT = {
  teenager: "少年",
  youth: "青年",
  "middle-aged": "中年",
  elder: "老年",
};

export default function TableAddonExample() {
  // 年龄段筛选值
  const [stage, setStage] = useState("all");
  // 婚姻状态筛选值
  const [marrages, setMarrages] = useState([]);
  // 当前排序列
  const [sorts, setSorts] = useState([]);
  // 当前选中行
  const [selected, setSelected] = useState(null);

  let filteredRecords = records.slice();

  // 根据年龄段筛选
  if (stage !== "all") {
    filteredRecords = filteredRecords.filter(x => x.stage === stage);
  }

  // 根据混音状态
  if (marrages.length > 0) {
    filteredRecords = filteredRecords.filter(
      record => marrages.indexOf(record.marrage) > -1
    );
  }

  // 如果要在前端排序，可以用 sortable.comparer 生成默认的排序方法
  filteredRecords.sort(sortable.comparer(sorts));

  return (
    <Table
      records={filteredRecords}
      recordKey={x => x.name}
      rowDisabled={record => record.name === "张三"}
      columns={[
        { key: "name", header: "姓名" },
        { key: "age", header: "年龄" },
        {
          key: "stage",
          header: "年龄段",
          render: x => STAGE_TEXT[x.stage],
        },
        {
          key: "marrage",
          header: "婚姻状态",
          render: x => MARRAGE_TEXT[x.marrage],
        },
      ]}
      addons={[
        // 支持表格滚动，高度超过 192 开始显示滚动条
        scrollable({
          maxHeight: 192,
          minWidth: 1600,
          onScrollBottom: () => console.log("到达底部"),
        }),

        // 对 age 列增加单选过滤支持
        filterable({
          type: "single",
          column: "stage",
          value: stage,
          onChange: value => setStage(value),
          // 增加 "全部" 选项
          all: {
            value: "all",
            text: "全部",
          },
          // 选项列表
          options: [
            { value: "teenager", text: "少年" },
            { value: "youth", text: "青年" },
            { value: "middle-aged", text: "中年" },
            { value: "elder", text: "老年" },
          ],
        }),

        // 对 marrage 列增加多选过滤支持
        filterable({
          type: "multiple",
          column: "marrage",
          value: marrages.map(x => String(x)),
          onChange: value => {
            setMarrages(value.filter(x => x !== "all").map(x => Number(x)));
          },
          all: {
            value: "all",
            text: "全部",
          },
          options: MARRAGE_TEXT.map((text, index) => ({
            value: index.toString(),
            text,
          })),
        }),

        // 支持表格排序
        sortable({
          // 这两列支持排序，其中 age 列优先倒序
          columns: ["name", { key: "age", prefer: "desc" }],
          value: sorts,
          onChange: value => setSorts(value),
        }),

        // 支持表格单选
        radioable({
          value: selected,
          rowSelect: true,
          onChange: (value, context) => {
            console.log(value, context);
            setSelected(value);
          },
          render: (element, { disabled }) => {
            return disabled ? <Icon type="loading" /> : element;
          },
        }),

        // 支持 props 注入
        injectable({
          row: () => ({
            onClick: () => console.log("Row Click"),
          }),
        }),
      ]}
    />
  );
}
