// @tencent/tea-component/src/blank/Blank.tsx
import React from "react";
import classNames from "classnames";
import { BlankSteps } from "./BlankSteps";
import { MediaObject } from "../mediaobject";
import { StyledProps } from "../_type/StyledProps";
import { withStatics } from "../_util/with-statics";

export interface BlankTheme {
  image: { background: string; gif: string };
}

export interface BlankProps extends StyledProps {
  /**
   * 预置主题
   */
  theme?: "error" | "open" | "arrears" | "permission" | BlankTheme;

  /**
   * 描述
   */
  description?: string;

  /**
   * 操作区
   */
  operation?: React.ReactNode;

  /**
   * 操作区下方内容
   */
  extra?: React.ReactNode;

  /**
   * 描述与操作区间内容
   */
  children?: React.ReactNode;
}

export const Blank = withStatics(
  function Blank({
    theme,
    description,
    operation,
    extra,
    children,
    className,
    style,
  }: BlankProps) {
    const { image } = getCustomTheme(theme);
    return (
      <div
        className={classNames(
          "tea-blank-page",
          {
            "default-error": theme === "error",
            "default-open": theme === "open",
            "default-arrears": theme === "arrears",
            "default-permission": theme === "permission",
          },
          className
        )}
        style={style}
      >
        <MediaObject
          align="middle"
          media={
            <div className="tea-blank-page__media">
              <span className="tea-blank-page__media-bg">
                {image.background && <img src={image.background} alt="" />}
              </span>
              <span className="tea-blank-page__media-gif">
                {image.gif && <img src={image.gif} alt="" />}
              </span>
            </div>
          }
        >
          <h4 className="tea-h4 tea-mb-2n">{description}</h4>
          {children}
          {(operation || extra) && (
            <div className="tea-blank-page__op">
              {operation}
              {extra && <div className="tea-mt-4n">{extra}</div>}
            </div>
          )}
        </MediaObject>
      </div>
    );
  },
  {
    Steps: BlankSteps,
  }
);

function getCustomTheme(theme: BlankProps["theme"]): BlankTheme {
  if (typeof theme === "string") {
    return { image: { background: null, gif: null } };
  }
  return {
    image: {},
    ...theme,
  };
}
