// @tencent/tea-component/lib/calendar/CalendarPart.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var classnames_1 = tslib_1.__importDefault(require("classnames"));
var justify_1 = require("../justify");
exports.default = {
    Panel: CalendarPanel,
    Header: CalendarHeader,
    Body: CalendarBody,
    Footer: CalendarFooter,
};
function CalendarPanel(_a) {
    var style = _a.style, className = _a.className, children = _a.children, _b = _a.rangeMode, rangeMode = _b === void 0 ? false : _b, _c = _a.timeMode, timeMode = _c === void 0 ? false : _c;
    return (react_1.default.createElement("div", { className: classnames_1.default("tea-calendar", className, {
            "tea-calendar--date-range": rangeMode,
            "tea-calendar--time": timeMode,
        }), style: style }, children));
}
exports.CalendarPanel = CalendarPanel;
function CalendarHeader(_a) {
    var style = _a.style, className = _a.className, children = _a.children;
    return (react_1.default.createElement("div", { className: classnames_1.default("tea-calendar__header", className), style: style }, children));
}
exports.CalendarHeader = CalendarHeader;
function CalendarBody(_a) {
    var style = _a.style, className = _a.className, children = _a.children;
    return (react_1.default.createElement("div", { className: classnames_1.default("tea-calendar__body", className), style: style }, children));
}
exports.CalendarBody = CalendarBody;
function CalendarFooter(_a) {
    var style = _a.style, className = _a.className, left = _a.left, right = _a.right;
    return (react_1.default.createElement("div", { className: classnames_1.default("tea-calendar__footer", className), style: style },
        react_1.default.createElement(justify_1.Justify, { left: left, right: right })));
}
exports.CalendarFooter = CalendarFooter;
