// @tencent/tea2-component/lib/grid/Grid.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var classnames_1 = tslib_1.__importDefault(require("classnames"));
var config_context_1 = require("../_util/config-context");
var GridContext = react_1.default.createContext({ gap: 20 });
function Row(_a) {
    var _b;
    var gap = _a.gap, verticalAlign = _a.verticalAlign, showSplitLine = _a.showSplitLine, children = _a.children, className = _a.className, style = _a.style;
    var classPrefix = config_context_1.useConfig().classPrefix;
    var grid = null;
    var rowStyle = null;
    // 定义了 gap 的，才生成一个 grid 上下文，否则使用默认样式即可
    if (typeof gap === "number") {
        grid = { gap: gap };
        rowStyle = {
            marginLeft: -gap / 2,
            marginRight: -gap / 2,
        };
    }
    rowStyle = tslib_1.__assign(tslib_1.__assign({}, (rowStyle || {})), (style || {}));
    var rowClassName = classnames_1.default(classPrefix + "-grid", className, (_b = {},
        _b[classPrefix + "-grid--split-line"] = showSplitLine,
        _b[classPrefix + "-vertical--" + verticalAlign] = verticalAlign,
        _b));
    return (react_1.default.createElement("div", { className: rowClassName, style: rowStyle },
        react_1.default.createElement(GridContext.Provider, { value: grid }, children)));
}
exports.Row = Row;
function Col(_a) {
    var span = _a.span, className = _a.className, style = _a.style, children = _a.children;
    var classPrefix = config_context_1.useConfig().classPrefix;
    var grid = react_1.useContext(GridContext);
    var colStyle = null;
    if (grid) {
        colStyle = {
            paddingLeft: grid.gap / 2,
            paddingRight: grid.gap / 2,
        };
    }
    var colClassName = classnames_1.default(classPrefix + "-grid__item" + (span ? "-" + span : ""), className);
    colStyle = tslib_1.__assign(tslib_1.__assign({}, (colStyle || {})), (style || {}));
    return (react_1.default.createElement("div", { className: colClassName, style: colStyle },
        react_1.default.createElement("div", { className: classPrefix + "-grid__box" }, children)));
}
exports.Col = Col;
