// @tencent/tea-component/src/select/_example/GroupSelectExample.jsx
import React, { useState } from "react";
import { Select } from "@tea/component/select";
import { Form } from "@tea/component/form";

const options = [
  { groupKey: "fruit", value: "strawberry", text: "草莓", tooltip: "甜甜甜" },
  {
    groupKey: "fruit",
    value: "apple",
    text: "苹果",
    disabled: true,
    tooltip: "每日一苹果，医生远离我",
  },
  { groupKey: "fruit", value: "orange", text: "橙子", tooltip: "丰富 VC 含量" },
  { groupKey: "drink", value: "coca-cola", text: "可口可乐" },
  {
    groupKey: "drink",
    value: "pepsi-cola",
    text: "百事可乐",
    tooltip: "百事可乐为什么比可口可乐好喝？",
  },
];

const groups = {
  fruit: "水果",
  drink: "饮料",
};

export default function SelectExample() {
  const [favorite, setFavorite] = useState(null);

  return (
    <Form>
      <Form.Item label="原生下拉">
        <Select
          type="native"
          size="m"
          groups={groups}
          options={options}
          value={favorite}
          onChange={value => setFavorite(value)}
          placeholder="请选择你最爱的水果"
        />
      </Form.Item>
      <Form.Item label="模拟下拉">
        <Select
          type="simulate"
          appearence="default"
          groups={groups}
          options={options}
          value={favorite}
          onChange={value => setFavorite(value)}
          placeholder="请选择你最爱的食品"
        />
      </Form.Item>
    </Form>
  );
}
