// @tencent/tea-component/src/collapse/Collapse.tsx
import React from "react";
import classNames from "classnames";
import { CollapsePanel } from "./CollapsePanel";
import { CollapseContext } from "./CollapseContext";
import { useDefault } from "../_util/use-default";
import { StyledProps } from "../_type";

export interface CollapseProps extends StyledProps {
  /**
   * 当前激活的面板 ID 组
   */
  activeIds?: string[];

  /**
   * 默认激活的面板 ID 组
   */
  defaultActiveIds?: string[];

  /**
   * 面板激活变化回调
   */
  onActive?: (
    activeIds: string[],
    context: { event: React.SyntheticEvent }
  ) => void;

  /**
   * 面板 <Collapse.Panel />
   */
  children?: React.ReactNode;
}

export function Collapse({
  activeIds,
  defaultActiveIds = [],
  onActive,
  children,
  className,
  style,
}: CollapseProps) {
  const [value, onChange] = useDefault(activeIds, defaultActiveIds, onActive);

  return (
    <div className={classNames("tea-accordion-group", className)} style={style}>
      <CollapseContext.Provider
        value={{
          activeIds: value,
          onActive: onChange,
        }}
      >
        {children}
      </CollapseContext.Provider>
    </div>
  );
}

Collapse.Panel = CollapsePanel;
