// @tencent/tea2-app/lib/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
// 导出国际化模块，以及其类型
var i18n_1 = require("./core/i18n");
exports.i18n = i18n_1.i18n;
// 导出 history，支持 React Router
var history_context_1 = require("./core/history/history-context");
exports.useHistory = history_context_1.useHistory;
exports.withHistory = history_context_1.withHistory;
// 导出 useBeforeDestroy hooks
var beforeDestroy_1 = require("./core/beforeDestroy");
exports.useBeforeDestroy = beforeDestroy_1.useBeforeDestroy;
// 导出 useDocumentTitle Hooks
var title_1 = require("./bridge/title");
exports.useDocumentTitle = title_1.useDocumentTitle;
// 导出 SDKLoader
var SDKLoader_1 = require("./bridge/sdk/SDKLoader");
exports.SDKLoader = SDKLoader_1.SDKLoader;
// 导出 app 命名空间
var app = __importStar(require("./core/app"));
exports.app = app;
