// @tencent/tea2-component/src/menu/_example/MenuExample.jsx
import React, { useState } from "react";
import { Layout } from "@tencent/tea-component/lib/layout";
import { Card } from "@tencent/tea-component/lib/card";
import { Text } from "@tencent/tea-component/lib/text";
import { ExternalLink } from "@tencent/tea-component/lib/link";
import { Menu } from "@tencent/tea-component/lib/menu";
import { Icon } from "@tencent/tea-component/lib/icon";

const { Body, Sider, Content } = Layout;

function MenuExample() {
  const [selected, setSelected] = useState("overview");

  const getMenuItemProps = id => ({
    selected: selected === id,
    onClick: () => setSelected(id),
  });

  return (
    <Layout>
      <Body>
        <Sider>
          <Menu
            back={{
              title: "产品名称产品名称产品名称",
              description: "数据测试",
              tooltip: "产品名称产品名称产品名称 - 数据测试勿删",
              onClick: () => console.log("back"),
            }}
          >
            <Menu.Item title="一级菜单" {...getMenuItemProps("1")} />
            <Menu.SubMenu defaultOpened title="一级菜单">
              <Menu.Item title="二级菜单" {...getMenuItemProps("2-1")} />
              <Menu.Item title="二级菜单" {...getMenuItemProps("2-2")} />
            </Menu.SubMenu>
            <Menu.Group title="菜单分组">
              <Menu.Item
                title={
                  <>
                    一级菜单
                    <Icon type="externallink" />
                  </>
                }
                {...getMenuItemProps("3")}
              />
            </Menu.Group>

            <Menu.Group title="菜单分组">
              <Menu.SubMenu title="一级菜单">
                <Menu.Item title="二级菜单" {...getMenuItemProps("4-1")} />
                <Menu.Item title="二级菜单" {...getMenuItemProps("4-2")} />
              </Menu.SubMenu>
            </Menu.Group>
          </Menu>
        </Sider>
        <Content>
          <Content.Header
            title="内容标题"
            subtitle={
              <>
                说明文字 <Text theme="label">带颜色说明文字</Text>
              </>
            }
            operation={<ExternalLink weak>内容帮助</ExternalLink>}
          />
          <Content.Body>
            {/* 内容区域一般使用 Card 组件显示内容 */}
            <Card>
              <Card.Body>内容卡片</Card.Body>
            </Card>
          </Content.Body>
        </Content>
      </Body>
    </Layout>
  );
}

export default function Demo() {
  return (
    <section
      style={{
        height: 600,
        border: "1px solid #ddd",
      }}
    >
      <MenuExample />
    </section>
  );
}
