// @tencent/tea-component/lib/_chart/core/event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var d3_1 = require("d3");
var EventListener = /** @class */ (function () {
    function EventListener(ctx) {
        this.ctx = ctx;
        var _a = this.ctx.plot, plot = _a.selection, width = _a.width, height = _a.height;
        this.selection = plot
            .append("rect")
            .style("height", height)
            .style("width", width)
            .attr("fill", "transparent")
            .attr("pointer-events", "all");
        this.start();
    }
    /**
     * 开始监听事件
     */
    EventListener.prototype.start = function () {
        if ("ontouchstart" in document) {
            this.selection
                .style("-webkit-tap-highlight-color", "transparent")
                .on("touchmove", this.plotMoved.bind(this))
                .on("touchstart", this.plotEntered.bind(this))
                .on("touchend", this.plotLeft.bind(this));
        }
        else {
            this.selection
                .on("mousemove", this.plotMoved.bind(this))
                .on("mouseenter", this.plotEntered.bind(this))
                .on("mouseleave", this.plotLeft.bind(this));
        }
    };
    /**
     * 监测鼠标移动
     */
    EventListener.prototype.plotMoved = function () {
        d3_1.event.preventDefault();
        this.ctx.emit("plot.moved", d3_1.event.layerX, d3_1.event.layerY);
    };
    /**
     * 监测鼠标进入绘图区
     */
    EventListener.prototype.plotEntered = function () {
        this.ctx.emit("plot.entered");
    };
    /**
     * 监测鼠标离开绘图区
     */
    EventListener.prototype.plotLeft = function () {
        this.ctx.emit("plot.left");
    };
    return EventListener;
}());
exports.EventListener = EventListener;
