// @tencent/tea-component/lib/select/SelectMultiple.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var dropdown_1 = require("../dropdown");
var list_1 = require("../list");
var checkbox_1 = require("../checkbox");
var form_1 = require("../form");
var checktree_1 = require("../checktree");
var button_1 = require("../button");
var i18n_1 = require("../i18n");
var with_statics_1 = require("../_util/with-statics");
var tips_1 = require("../tips");
var searchbox_1 = require("../searchbox");
var noop = function () { return null; };
exports.SelectMultiple = with_statics_1.withStatics(function SelectMultiple(props) {
    var e_1, _a;
    var _b = form_1.useDefaultValue(props, []), _c = _b.staging, staging = _c === void 0 ? true : _c, value = _b.value, onChange = _b.onChange, _d = _b.options, options = _d === void 0 ? [] : _d, allOption = _b.allOption, _e = _b.shouldOptionExcludeFromAll, shouldOptionExcludeFromAll = _e === void 0 ? "disabled" : _e, _f = _b.allowEmpty, allowEmpty = _f === void 0 ? true : _f, searchable = _b.searchable, _g = _b.searchPlaceholder, searchPlaceholder = _g === void 0 ? "" : _g, _h = _b.onSearch, onSearch = _h === void 0 ? noop : _h, _j = _b.filter, filter = _j === void 0 ? function (inputValue, _a) {
        var text = _a.text, value = _a.value;
        var optionText = String(typeof text === "string" ? text : value);
        return !searchable || optionText.includes(inputValue);
    } : _j, dropdownProps = tslib_1.__rest(_b, ["staging", "value", "onChange", "options", "allOption", "shouldOptionExcludeFromAll", "allowEmpty", "searchable", "searchPlaceholder", "onSearch", "filter"]);
    var t = i18n_1.useTranslation();
    var _k = tslib_1.__read(react_1.useState(value), 2), stagingValue = _k[0], setStagingValue = _k[1];
    var InputRef = react_1.useRef(null);
    var _l = tslib_1.__read(react_1.useState(""), 2), inputValue = _l[0], setInputValue = _l[1];
    function focus() {
        if (searchable) {
            setInputValue("");
            onSearch("");
            setTimeout(function () {
                if (InputRef.current) {
                    InputRef.current.focus();
                }
            }, 100); // 第一次展开时 Input 还未渲染
        }
    }
    var handleChange = function (value, context) {
        if (staging === false) {
            onChange(value, context);
            return;
        }
        setStagingValue(value);
    };
    // 「全部」选项和其余选项是数关系
    var releations = {};
    if (allOption) {
        try {
            for (var options_1 = tslib_1.__values(options), options_1_1 = options_1.next(); !options_1_1.done; options_1_1 = options_1.next()) {
                var option = options_1_1.value;
                if (
                // 用户可以指定哪些选项从全选逻辑中排除
                (typeof shouldOptionExcludeFromAll === "function" &&
                    shouldOptionExcludeFromAll(option)) ||
                    (shouldOptionExcludeFromAll === "disabled" && option.disabled)) {
                    // continue
                }
                else {
                    releations[option.value] = allOption.value;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (options_1_1 && !options_1_1.done && (_a = options_1.return)) _a.call(options_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    }
    // 筛选
    var filteredOptions = options.filter(function (options) {
        return filter(inputValue, options);
    });
    var tips = props.tips;
    if (!tips && filteredOptions.length === 0) {
        tips = react_1.default.createElement(tips_1.EmptyTip, null);
    }
    tips = tips ? react_1.default.createElement(list_1.List.StatusTip, null, tips) : null;
    return (react_1.default.createElement(dropdown_1.Dropdown, tslib_1.__assign({ updateOnChildrenChange: true, button: react_1.default.createElement(ValueBrief, tslib_1.__assign({}, props, { value: value })), onOpen: function () {
            setStagingValue(value);
            focus();
        }, clickClose: false }, dropdownProps), function (close) { return (react_1.default.createElement(checktree_1.CheckTree, { relations: releations, value: (staging ? stagingValue : value) || value, onChange: handleChange, disabledNames: options.filter(function (x) { return x.disabled; }).map(function (x) { return x.value; }) },
        searchable && (react_1.default.createElement(searchbox_1.SearchBox, { simple: true, ref: InputRef, value: inputValue, onChange: function (value) {
                setInputValue(value);
                onSearch(value);
            }, onClear: focus, placeholder: searchPlaceholder })),
        react_1.default.createElement(list_1.List, { type: "option", className: "tea-list--checkoption" },
            allOption &&
                options.length === filteredOptions.length &&
                renderSelectOption(allOption),
            tips,
            filteredOptions.map(function (option) { return renderSelectOption(option); })),
        staging !== false && (react_1.default.createElement(dropdown_1.Dropdown.Footer, null,
            react_1.default.createElement(button_1.Button, { type: "primary", disabled: !allowEmpty && stagingValue.length === 0, onClick: function (event) {
                    onChange(stagingValue, { event: event });
                    close();
                } }, t.okText),
            react_1.default.createElement(button_1.Button, { type: "weak", onClick: close }, t.cancelText))))); }));
}, {
    defaultLabelAlign: "middle",
});
function renderSelectOption(option) {
    return (react_1.default.createElement(list_1.List.Item, { key: option.value, tooltip: option.tooltip, disabled: option.disabled },
        react_1.default.createElement(checkbox_1.Checkbox, { onClick: function (evt) { return evt.stopPropagation(); }, name: option.value }, option.text || option.value)));
}
function ValueBrief(_a) {
    var value = _a.value, allOption = _a.allOption, options = _a.options, button = _a.button, placeholder = _a.placeholder;
    var t = i18n_1.useTranslation();
    if (button) {
        return react_1.default.createElement(react_1.Fragment, null, button);
    }
    if (!value || !value.length) {
        return react_1.default.createElement(react_1.Fragment, null, placeholder || t.pleaseSelect);
    }
    if (allOption && value.length === options.length) {
        return react_1.default.createElement(react_1.Fragment, null, allOption.text || allOption.value);
    }
    var showupOption = value
        .slice(0, 5)
        .map(function (x) { return options.find(function (y) { return y.value === x; }); });
    var showupExcceed = showupOption.length < value.length;
    return (react_1.default.createElement(react_1.Fragment, null,
        showupOption
            .filter(function (x) { return x; })
            .map(function (option, index) { return (react_1.default.createElement(react_1.Fragment, { key: option.value },
            option.text || option.value,
            index < showupOption.length - 1 && ", ")); }),
        showupExcceed && "..."));
}
